### ISSUE TITLE
Racing substituers instead of downloading exhaustively

### ISSUE THREAD

**Issue by masaeedu** [NixOS/nix#3281]
I have a Nix daemon set up on my non-NixOS system Arch box, and have my user set up as a trusted user. In my `~/.config/nix/nix.conf` I have a large number of substituters defined:

```
show-trace = true
substituters = https://cache.nixos.org https://mcneissue.cachix.org https://cache.dhall-lang.org https://all-hies.cachix.org https://nixfmt.cachix.org https://hercules-ci.cachix.org https://masaeedu.cachix.org https://nix-tools.cachix.org
trusted-public-keys = cache.nixos.org-1:6NCHdD59X431o0gWypbMrAURkbJ16ZPMQFGspcDShjY= mcneissue.cachix.org-1:IFYZ9RUtUdFm23VPyl8k7dpv6i8GzeslfeASKbZeGTg= cache.dhall-lang.org:I9/H18WHd60olG5GsIjolp7CtepSgJmM2CsO813VTmM= all-hies.cachix.org-1:JjrzAOEUsD9ZMt8fdFbzo3jNAyEWlPAwdVuHw4RD43k= nixfmt.cachix.org-1:uyEQg16IhCFeDpFV07aL+Dbmh18XHVUqpkk/35WAgJI= hercules-ci.cachix.org-1:ZZeDl9Va+xe9j+KqdzoBZMFJHVQ42Uu/c/1/KMC5Lw0= masaeedu.cachix.org-1:3aZ9AliNA0dlqijZb91jWUViym2L32enVDEQd/uGg/s= nix-tools.cachix.org-1:ebBEBZLogLxcCvipq2MTvuHlP7ZRdkazFSQsbs0Px1A=
```

The behavior I expect from defining all these substituters is that when Nix tries to reify a derivation, it make requests to all configured substituters in parallel, and begin downloading from the first one (if any) that actually responds with a narinfo file.

Instead, what seems to happen is that whenever Nix wants to reify some derivation, it slowly walks down the list of substituters, making a request to each one sequentially. The more substituters I add, the slower all my builds become:

```
querying info about '/nix/store/axj67jai1kmcl4nh5pf7l3i118qr1bdg-git-ls-files' on 'https://cache.dhall-lang.org'...
downloading 'https://cache.dhall-lang.org/axj67jai1kmcl4nh5pf7l3i118qr1bdg.narinfo'...
querying info about '/nix/store/axj67jai1kmcl4nh5pf7l3i118qr1bdg-git-ls-files' on 'https://cache.nixos.org'...
downloading 'https://cache.nixos.org/axj67jai1kmcl4nh5pf7l3i118qr1bdg.narinfo'...
querying info about '/nix/store/axj67jai1kmcl4nh5pf7l3i118qr1bdg-git-ls-files' on 'https://mcneissue.cachix.org'...
downloading 'https://mcneissue.cachix.org/axj67jai1kmcl4nh5pf7l3i118qr1bdg.narinfo'...
querying info about '/nix/store/axj67jai1kmcl4nh5pf7l3i118qr1bdg-git-ls-files' on 'https://all-hies.cachix.org'...
downloading 'https://all-hies.cachix.org/axj67jai1kmcl4nh5pf7l3i118qr1bdg.narinfo'...
querying info about '/nix/store/axj67jai1kmcl4nh5pf7l3i118qr1bdg-git-ls-files' on 'https://nixfmt.cachix.org'...
downloading 'https://nixfmt.cachix.org/axj67jai1kmcl4nh5pf7l3i118qr1bdg.narinfo'...
querying info about '/nix/store/axj67jai1kmcl4nh5pf7l3i118qr1bdg-git-ls-files' on 'https://hercules-ci.cachix.org'...
downloading 'https://hercules-ci.cachix.org/axj67jai1kmcl4nh5pf7l3i118qr1bdg.narinfo'...
querying info about '/nix/store/axj67jai1kmcl4nh5pf7l3i118qr1bdg-git-ls-files' on 'https://masaeedu.cachix.org'...
downloading 'https://masaeedu.cachix.org/axj67jai1kmcl4nh5pf7l3i118qr1bdg.narinfo'...
querying info about '/nix/store/axj67jai1kmcl4nh5pf7l3i118qr1bdg-git-ls-files' on 'https://nix-tools.cachix.org'...
downloading 'https://nix-tools.cachix.org/axj67jai1kmcl4nh5pf7l3i118qr1bdg.narinfo'...
... etc.
```

Am I misunderstanding how substituters are supposed to work or misconfiguring something?

**Comment by edolstra** [comment]
Yes, substituters are tried sequentially at the moment.

**Comment by michaelpj** [comment]
See also https://github.com/NixOS/nix/issues/3019

**Comment by domenkozar** [comment]
I am considering batching those with something like https://cachix.org/caches/mcneissue/all-hies/axj67jai1kmcl4nh5pf7l3i118qr1bdg.narinfo

That only partially solves the problem for cachix bits.

Cachix also uses a mass query, you give it a list of paths and it will return a list of paths that are missing. Something similar could be added to the binary cache api in Nix.

**Comment by masaeedu** [comment]
@michaelpj Would you say this is a duplicate of #3019? If so, we can close this one.

**Comment by michaelpj** [comment]
I think this is a duplicate, but I don't really mind which one we keep.
